datumaro/components/validator.py

```diff
@@ -0,0 +1,951 @@
+# Copyright (C) 2021 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+from copy import deepcopy
+from enum import Enum
+import numpy as np
+
+from datumaro.components.dataset import IDataset
+from datumaro.components.extractor import AnnotationType
+
+
+Severity = Enum('Severity',
+    [
+        'warning',
+        'error'
+    ]
+)
+
+
+class _Validator:
+    """
+    A base class for task-specific validators.
+
+    ...
+
+    Attributes
+    ----------
+    task_type : str
+        task type (ie. 'classification', etc.) (default is 'classification')
+    ann_type : AnnotationType
+        annotation type to validate (default is AnnotationType.label)
+    far_from_mean_thr : float
+        constant used to define mean +/- k * stdev (default is None)
+
+    Methods
+    -------
+    compute_statistics(dataset):
+        Computes various statistics of the dataset based on task type.
+    generate_reports(stats):
+        Abstract method that must be implemented in a subclass.
+    """
+
+    def __init__(self,
+                task_type='classification',
+                ann_type=AnnotationType.label,
+                far_from_mean_thr=None):
+        self.task_type = task_type
+        self.ann_type = ann_type
+        self.far_from_mean_thr = far_from_mean_thr
+
+    def compute_statistics(self, dataset):
```

Comment:
Please try to split this function into smaller parts. Maybe, using local functions.